### PR TITLE
[jsk_pcl_ros]Add color based region

### PR DESCRIPTION
**Pull request by otsubo**
add color based region growing segmentation nodelet
http://pointclouds.org/documentation/tutorials/region_growing_rgb_segmentation.php
screenshot_from_2016-03-26 21 27 06
screenshot_from_2016-03-26 21 27 34

### CODE REVIEW

File: jsk_pcl_ros/cfg/ColorBasedRegionGrowingSegmentation.cfg [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python
```
Comment (orikuma) [inline jsk_pcl_ros/cfg/ColorBasedRegionGrowingSegmentation.cfg]: Please write Software License Agreement as header.

File: jsk_pcl_ros/src/color_based_region_growing_segmentation_nodelet.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  void ColorBasedRegionGrowingSegmentation::segment(const sensor_msgs::PointCloud2::ConstPtr& msg)
+  {
+    boost::mutex::scoped_lock lock(mutex_);
+    pcl::search::Search <pcl::PointXYZRGB>::Ptr tree = boost::shared_ptr<pcl::search::Search<pcl::PointXYZRGB> > (new pcl::search::KdTree<pcl::PointXYZRGB>);
+    pcl::PointCloud <pcl::PointXYZRGB>::Ptr cloud (new pcl::PointCloud <pcl::PointXYZRGB>);
+    pcl::fromROSMsg(*msg, *cloud);
+
+    pcl::IndicesPtr indices (new std::vector <int>);
```
Comment (orikuma) [inline jsk_pcl_ros/src/color_based_region_growing_segmentation_nodelet.cpp]: PassThroughFilter should be applied outside of this node (pass to other nodelet etc)

File: jsk_pcl_ros/include/jsk_pcl_ros/color_based_region_growing_segmentation.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
+// -*- mode: C++ -*-
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2013, Ryohei Ueda and JSK Lab
```
Comment (orikuma) [inline jsk_pcl_ros/include/jsk_pcl_ros/color_based_region_growing_segmentation.h]: Copyright is "Copyright (c) 2016, Satoshi Otsubo and JSK Lab"

File: jsk_pcl_ros/src/color_based_region_growing_segmentation_nodelet.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    if (min_cluster_size_ != config.min_cluster_size) {
+      min_cluster_size_ = config.min_cluster_size;
+    }
+  }
+
+  void ColorBasedRegionGrowingSegmentation::segment(const sensor_msgs::PointCloud2::ConstPtr& msg)
+  {
+    boost::mutex::scoped_lock lock(mutex_);
+    pcl::search::Search <pcl::PointXYZRGB>::Ptr tree = boost::shared_ptr<pcl::search::Search<pcl::PointXYZRGB> > (new pcl::search::KdTree<pcl::PointXYZRGB>);
```
Comment (orikuma) [inline jsk_pcl_ros/src/color_based_region_growing_segmentation_nodelet.cpp]: pcl::search::Search <pcl::PointXYZRGB>::Ptr is better than boost::shared_ptr<pcl::search::Search<pcl::PointXYZRGB>>

File: jsk_pcl_ros/src/color_based_region_growing_segmentation_nodelet.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
+#include <pcl_conversions/pcl_conversions.h>
```
Comment (orikuma) [inline jsk_pcl_ros/src/color_based_region_growing_segmentation_nodelet.cpp]: Please write Software License Agreement as header.

**Comment by k-okada** [comment]
@otsubo did you add document for this?

**Comment by otsubo** [comment]
@k-okada I have added documentation for this in https://github.com/jsk-ros-pkg/jsk_recognition/pull/1698

**Comment by wkentaro** [comment]
(I don't think you need split the PRs... but it's ok.)

Could you please add some examples and test?
There are already test files to publish input point cloud for the node, so you can just use it.
https://github.com/jsk-ros-pkg/jsk_recognition/tree/master/jsk_pcl_ros

**Comment by k-okada** [comment]
ping @otsubo

**Comment by k-okada** [comment]
@otsubo I tried to creat test code as follows but it did not run, how did you test this program?
```
<launch>

  <!-- load log -->
  <rosparam command="load"
            file="$(find jsk_pcl_ros)/test_data/2015-11-04-19-37-29_baxter-kiva-object-in-hand-cloud/rosparam.yaml" />
  <node name="rosbag_play"
        pkg="rosbag" type="play"
        args="$(find jsk_pcl_ros)/test_data/2015-11-04-19-37-29_baxter-kiva-object-in-hand-cloud/vision.compressed.bag --clock --loop">
  </node>

  <arg name="INPUT_CLOUD" value="/kinect2_head/hd/points" />

  <node name="color_based_region_growing_segmentation_nodelet"
        pkg="nodelet" type="nodelet"
        args="standalone jsk_pcl/RegionGrowingSegmentation">
    <remap from="~input" to="$(arg INPUT_CLOUD)" />
  </node>
</launch>
```

## ERROR

```
Failed to find match for field 'normal_x'.
Failed to find match for field 'normal_y'.
Failed to find match for field 'normal_z'.
Failed to find match for field 'curvature'.
nodelet: /build/pcl-1.7-K_Z193/pcl-1.7-1.7.1/kdtree/include/pcl/kdtree/impl/kdtree_flann.hpp:136: int pcl::KdTreeFLANN<PointT, Dist>::nearestKSearch(const PointT&, int, std::vector<int>&, std::vector<float>&) const [with PointT = pcl::PointNormal; Dist = flann::L2_Simple<float>]: Assertion `point_representation_->isValid (point) && "Invalid (NaN, Inf) point coordinates given to nearestKSearch!"' failed.
[color_based_region_growing_segmentation_nodelet-2] process has died [pid 32554, exit code -6, cmd /opt/ros/indigo/lib/nodelet/nodelet standalone jsk_pcl/RegionGrowingSegmentation ~input:=/kinect2_head/hd/points __name:=color_based_region_growing_segmentation_nodelet __log:=/home/k-okada/.ros/log/2d53144e-75a1-11e6-a61d-28b2bde39261/color_based_region_growing_segmentation_nodelet-2.log].
log file: /home/k-okada/.ros/log/2d53144e-75a1-11e6-a61d-28b2bde39261/color_based_region_growing_segmentation_nodelet-2*.log
```

## To get bag file
```
catkin -v -i bt --start-with-this --make-args install_test_data
```

**Comment by otsubo** [comment]
sorry for missing your ping.
I'll check it soon.

**Comment by k-okada** [comment]
```
        args="standalone jsk_pcl/RegionGrowingSegmentation">
```
was not correct, need to change
````
        args="standalone jsk_pcl/ColorRegionGrowingSegmentation">
```
and I have created test code at https://github.com/jsk-ros-pkg/jsk_recognition/pull/1866/commits/388e1de9f4cbff917ab14a005175c57712f8d91f, please check this.

BTW, this process is super slow, is this because of parameter settings? input data or algorithm itself? If you're using program slower than 5hz, I recommend not to use that, you can get bounding box of color region use https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_pcl_ros/src/color_filter_nodelet.cpp with vivid colors objects

**Comment by k-okada** [comment]
rewrited at #1866

**Comment by otsubo** [comment]
> BTW, this process is super slow, is this because of parameter settings? input data or algorithm itself?
It may be because of algorithm itself. 
I wanted to use region growing to find out end points of cable, but I don't need it now because of its speed.

>  you can get bounding box of color region use https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_pcl_ros/src/color_filter_nodelet.cpp with vivid colors objects
I understand. I'll use it when I need it.

**Pull request closed without merging**